lumicks/pylake/kymotracker/detail/peakfinding.py

```diff
@@ -97,13 +97,166 @@ def flatten(self):
+
+    This function generates indices for sampling based on the left-most and right-most bound. It
+    also generates appropriate pixel centers and weights that account for the fact that they are
+    only partial pixels.
+
+    Parameters
+    ----------
+    index_array: np.ndarray
+        2D pixel indices
+    left_edges, right_edges : np.ndarray
+        Lower and upper pixel edges
+
+    Returns
+    -------
+    selection : np.ndarray
+        Boolean mask with shape `(N, M)` which select which points from the window to use.
+        `N` is the number of pixels being refined, while `M` is the window size used for refinement.
+    centers : np.ndarray
+        Pixel center positions with shape `(N, M)` with edge pixel positions accounting for being
+        partial pixels. `N` is the number of pixels being refined, while `M` is the window size
+        used for refinement.
+    weights : np.ndarray
+        Weights with shape `(N, M)` that account for down-weighting the edges.
+        `N` is the number of pixels being refined, while `M` is the window size used for refinement.
+    """
+    left_floor = np.floor(left_edges).astype(int)
+    right_ceil = np.ceil(right_edges).astype(int)
+
+    # Upper bound needs to be rounded inwards
+    right_floor = np.floor(np.nextafter(right_edges, 0)).astype(int)
+
+    # Find which pixels we are still using
+    active = np.logical_and(
+        index_array >= left_floor[:, np.newaxis], index_array < right_ceil[:, np.newaxis]
+    )
+
+    num_points = left_floor.size
+    centers = active * (index_array + 0.5)
+
+    centers[np.arange(num_points), left_floor] = (left_edges + left_floor + 1) / 2
+    centers[np.arange(num_points), right_floor] = (right_edges + right_ceil - 1) / 2
+
+    # Pixels outside the bounds are not used
+    weights = active.astype(float)
+
+    # Fractional pixels are down-weighted.
+    weights[np.arange(num_points), left_floor] = 1.0 - (left_edges - np.floor(left_edges))
+    weights[np.arange(num_points), right_floor] = 1.0 - (np.ceil(right_edges) - right_edges)
+    return active, centers, weights
+
+
+def unbiased_centroid(position, data, tolerance=1e-3, max_iterations=50, epsilon=1e-8):
+    """Perform an unbiased centroid estimation in a vectorized manner
+
+    The bias in centroid refinement is proportional to the asymmetry around the spot. To remove this
+    bias, we define a virtual window that's symmetric around the spot position. In practice, this
+    involves removing fractional pixels at the edges. If we want to shift the center of the image
+    by 1 pixel, we need to move an edge point 2 pixels.
+
+    Fractional pixels are dealt with by down-weighting edge pixels and recomputing a new pixel
+    center for those pixels.
+
+    For more information on this method, please refer to [1]_.
+
+    Parameters
+    ----------
+    position : np.ndarray
+        Subpixel coordinates obtained with regular centroid method.
+    data : np.ndarray
+        2D array
```

Comment:
```suggestion
        2D image array
```

Reply:
👍

Reply:
I've rephrase it a bit for clarity (also because it took me a few seconds to figure out what it was again):
```
        2D image array with data used for refinement. It contains a small region around each spot.
        It has shape `(N, M)`, with `N` the number of pixels being refined, while `M` is the
        window size used for refinement.
```